In [9]:
from pathlib import Path
import pandas as pd
import numpy as np

PROJECT_ROOT = Path.home() / "Documents" / "spotify_project"

RAW_DATA = PROJECT_ROOT / "data" / "raw"
CLEAN_DATA = PROJECT_ROOT / "data" / "clean"
FIGURES = PROJECT_ROOT / "figures"

In [10]:
print("Project:", PROJECT_ROOT)
print("Raw data:", RAW_DATA)
print("Clean data:", CLEAN_DATA)

Project: /Users/jannoelvero/Documents/spotify_project
Raw data: /Users/jannoelvero/Documents/spotify_project/data/raw
Clean data: /Users/jannoelvero/Documents/spotify_project/data/clean


In [11]:
print(list(RAW_DATA.iterdir()))

[PosixPath('/Users/jannoelvero/Documents/spotify_project/data/raw/spotify_data_clean.csv'), PosixPath('/Users/jannoelvero/Documents/spotify_project/data/raw/.gitkeep'), PosixPath('/Users/jannoelvero/Documents/spotify_project/data/raw/track_data_final.csv')]


In [12]:
df_track = pd.read_csv(RAW_DATA / "track_data_final.csv")
df_spotify = pd.read_csv(RAW_DATA / "spotify_data_clean.csv")

In [13]:
print("Track dataset:", df_track.shape)
print("Spotify dataset:", df_spotify.shape)

Track dataset: (8778, 15)
Spotify dataset: (8582, 15)


In [14]:
df = df_track.copy()

In [15]:
df.shape

(8778, 15)

In [16]:
df = df[df["album_release_date"] != "0000"].copy()

In [17]:
df.shape

(8776, 15)

In [18]:
df.isnull().sum()

track_id              0
track_name            0
track_number          0
track_popularity      0
track_duration_ms     0
explicit              0
artist_name           4
artist_popularity     4
artist_followers      4
artist_genres         4
album_id              0
album_name            0
album_release_date    0
album_total_tracks    0
album_type            0
dtype: int64

In [19]:
df[df["artist_name"].isnull()][
    [
        "track_id",
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres"
    ]
]

,track_id,track_name,artist_name,artist_popularity,artist_followers,artist_genres
436,3QhfFRPkhPCR1RMJWV1gde,Radio,NaN,NaN,NaN,NaN
1396,2NbsBSshscKIPRG3kwBbfo,Never Felt So Alone,NaN,NaN,NaN,NaN
3885,5EhZC8GFOVXrpiUW22jncE,Urban Twilight,NaN,NaN,NaN,NaN
3886,0fM9dEhUFV4MHDuJgrcfOv,How You Get The Girl,NaN,NaN,NaN,NaN


In [20]:
missing_ids = df[df["artist_name"].isnull()]["track_id"]

df_spotify[
    df_spotify["track_id"].isin(missing_ids)
][
    [
        "track_id",
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres"
    ]
]

,track_id,track_name,artist_name,artist_popularity,artist_followers,artist_genres
1807,2NbsBSshscKIPRG3kwBbfo,Never Felt So Alone,NaN,0,0,NaN
5904,0fM9dEhUFV4MHDuJgrcfOv,How You Get The Girl,Taylor Swift,100,145396321,"country, pop, indie, folk"
6517,3QhfFRPkhPCR1RMJWV1gde,Radio,NaN,0,0,NaN
6679,5EhZC8GFOVXrpiUW22jncE,Urban Twilight,NaN,0,0,NaN


In [21]:
track_id = "0fM9dEhUFV4MHDuJgrcfOv"

df.loc[df["track_id"] == track_id, "artist_name"] = "Taylor Swift"
df.loc[df["track_id"] == track_id, "artist_popularity"] = 100
df.loc[df["track_id"] == track_id, "artist_followers"] = 145396321
df.loc[df["track_id"] == track_id, "artist_genres"] = "country, pop, indie, folk"

In [22]:
df[df["track_id"] == track_id][
    [
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres"
    ]
]

,track_name,artist_name,artist_popularity,artist_followers,artist_genres
3886,How You Get The Girl,Taylor Swift,100.0,145396321.0,"country, pop, indie, folk"


In [23]:
df[df["artist_name"].isnull()][
    [
        "track_id",
        "track_name",
        "album_name",
        "album_release_date",
        "track_popularity"
    ]
]

,track_id,track_name,album_name,album_release_date,track_popularity
436,3QhfFRPkhPCR1RMJWV1gde,Radio,Born To Die – Paradise Edition (Special Version),2012-01-01,65
1396,2NbsBSshscKIPRG3kwBbfo,Never Felt So Alone,Never Felt So Alone,2023-04-07,68
3885,5EhZC8GFOVXrpiUW22jncE,Urban Twilight,Darkbloom,2011-04-18,30


In [24]:
df[
    df["track_name"].isin(
        ["Radio", "Never Felt So Alone", "Urban Twilight"]
    )
][
    [
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres",
        "album_name"
    ]
]

,track_name,artist_name,artist_popularity,artist_followers,artist_genres,album_name
436,Radio,NaN,NaN,NaN,NaN,Born To Die – Paradise Edition (Special Version)
1038,Radio,Lana Del Rey,88.0,51773650.0,[],Born To Die (Deluxe Version)
1396,Never Felt So Alone,NaN,NaN,NaN,NaN,Never Felt So Alone
3885,Urban Twilight,NaN,NaN,NaN,NaN,Darkbloom
8759,Radio,Lana Del Rey,88.0,51847587.0,[],Born To Die - The Paradise Edition


In [25]:
df.loc[
    df["track_id"] == "3QhfFRPkhPCR1RMJWV1gde",
    "artist_name"
] = "Lana Del Rey"

In [26]:
df[df["track_id"] == "3QhfFRPkhPCR1RMJWV1gde"][
    ["track_name", "artist_name", "artist_popularity", "artist_followers", "artist_genres"]
]

,track_name,artist_name,artist_popularity,artist_followers,artist_genres
436,Radio,Lana Del Rey,NaN,NaN,NaN


In [27]:
df[
    df["album_name"] == "Never Felt So Alone"
][
    [
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres",
        "album_name"
    ]
]

,track_name,artist_name,artist_popularity,artist_followers,artist_genres,album_name
1396,Never Felt So Alone,NaN,NaN,NaN,NaN,Never Felt So Alone


In [28]:
df[
    df["track_name"] == "Urban Twilight"
][
    [
        "track_name",
        "artist_name",
        "artist_popularity",
        "artist_followers",
        "artist_genres",
        "album_name"
    ]
]

,track_name,artist_name,artist_popularity,artist_followers,artist_genres,album_name
3885,Urban Twilight,NaN,NaN,NaN,NaN,Darkbloom


In [29]:
df.isnull().sum()

track_id              0
track_name            0
track_number          0
track_popularity      0
track_duration_ms     0
explicit              0
artist_name           2
artist_popularity     3
artist_followers      3
artist_genres         3
album_id              0
album_name            0
album_release_date    0
album_total_tracks    0
album_type            0
dtype: int64

In [30]:
df["track_duration_min"] = (
    df["track_duration_ms"] / 60000
).round(2)

In [31]:
df = df.drop(columns=["track_duration_ms"])

In [32]:
df[["track_name", "track_duration_min"]].head()

,track_name,track_duration_min
0,3,3.55
1,Clouds,2.65
2,Forever & Always (Taylor’s Version),3.76
3,I Didn't Change My Number,2.64
4,Man Down,4.45


In [33]:
df["artist_genres"].head(20)

0                                               ['pop']
1                                     ['stutter house']
2                                                    []
3                                                    []
4                                                    []
5                                         ['dream pop']
6                                                    []
7     ['brazilian bass', 'electronic', 'slap house',...
8                                          ['soft pop']
9                                                    []
10                                              ['pop']
11           ['uk drill', 'drill', 'uk grime', 'grime']
12                                              ['pop']
13                                ['edm', 'electronic']
14                                                   []
15                                                   []
16                                                   []
17                    ['hip hop', 'west coast hi

In [34]:
df["artist_genres"].value_counts(dropna=False).head(20)

artist_genres
[]                                                                            4490
['soundtrack']                                                                 346
['pop']                                                                        303
['soft pop']                                                                   209
['rap']                                                                        192
['dark r&b']                                                                   111
['k-pop']                                                                       93
['grunge', 'rock']                                                              91
['country']                                                                     90
['edm']                                                                         82
['rap', 'hip hop']                                                              77
['country', 'acoustic country']                                          

In [35]:
df["artist_genres"] = df["artist_genres"].replace("[]", np.nan)

In [36]:
df["artist_genres"].isnull().sum()

np.int64(4493)

In [37]:
df["artist_genres"] = (
    df["artist_genres"]
    .str.replace("[", "", regex=False)
    .str.replace("]", "", regex=False)
    .str.replace("'", "", regex=False)
)

In [38]:
df["artist_genres"].value_counts().head(10)

artist_genres
soundtrack      346
pop             303
soft pop        209
rap             192
dark r&b        111
k-pop            93
grunge, rock     91
country          90
edm              82
rap, hip hop     77
Name: count, dtype: int64

In [39]:
df["release_year"] = pd.to_numeric(
    df["album_release_date"].str[:4],
    errors="coerce"
)

In [40]:
df[["album_release_date", "release_year"]].head(10)

,album_release_date,release_year
0,2009-11-09,2009
1,2023-01-13,2023
2,2021-04-09,2021
3,2021-07-30,2021
4,2010-01-01,2010
5,2015-08-28,2015
6,2017-12-22,2017
7,2024-09-27,2024
8,2014-06-21,2014
9,2020-07-20,2020


In [41]:
print("Earliest year:", df["release_year"].min())
print("Latest year:", df["release_year"].max())

Earliest year: 1952
Latest year: 2025


In [42]:
print("Missing release years:", df["release_year"].isnull().sum())

Missing release years: 0


In [43]:
df[
    [
        "track_popularity",
        "artist_popularity",
        "artist_followers",
        "album_total_tracks",
        "track_duration_min"
    ]
].describe()

,track_popularity,artist_popularity,artist_followers,album_total_tracks,track_duration_min
count,8776.000000,8773.000000,8.773000e+03,8776.000000,8776.000000
mean,52.244758,69.974353,2.437358e+07,13.782589,3.503447
std,24.065045,19.507944,3.816035e+07,11.797828,1.054251
min,0.000000,0.000000,0.000000e+00,1.000000,0.070000
25%,39.000000,60.000000,5.148960e+05,6.000000,2.900000
50%,58.000000,74.000000,6.294820e+06,13.000000,3.450000
75%,71.000000,84.000000,3.055055e+07,17.000000,4.000000
max,100.000000,100.000000,1.455421e+08,181.000000,13.520000


In [44]:
df[df["track_duration_min"] < 0.5][
    [
        "track_name",
        "artist_name",
        "track_duration_min",
        "track_popularity",
        "album_name"
    ]
].sort_values("track_duration_min")

,track_name,artist_name,track_duration_min,track_popularity,album_name
577,Up next: A few songs,Spotify,0.07,0,Daily Wellness
177,Dorothy's Interlude,Sam Smith,0.15,0,Gloria (Special Edition)
1332,EUROPA :),Bad Bunny,0.20,13,nadie sabe lo que va a pasar mañana
8115,I Can't Fucking Sing,The Weeknd,0.20,18,Hurry Up Tomorrow
5783,'Let's Just Sit',Marion McPherson,0.22,0,Lady Bird - Soundtrack from the Motion Picture
1990,!!!!!!!,Billie Eilish,0.23,17,"WHEN WE ALL FALL ASLEEP, WHERE DO WE GO?"
5518,Hurting Interlude,Sam Smith,0.31,0,Gloria (Special Edition)
5788,Until We're Skin & Bones,The Weeknd,0.37,21,Hurry Up Tomorrow
311,Paul Rosenberg - Skit,Eminem,0.38,10,The Eminem Show
7451,Her Majesty - Remastered 2009,The Beatles,0.43,27,Abbey Road (Remastered)


In [45]:
df[df["artist_name"] == "Spotify"][
    [
        "track_id",
        "track_name",
        "artist_name",
        "track_popularity",
        "artist_popularity",
        "artist_followers",
        "artist_genres",
        "album_name",
        "album_type",
        "track_duration_min"
    ]
]

,track_id,track_name,artist_name,track_popularity,artist_popularity,artist_followers,artist_genres,album_name,album_type,track_duration_min
577,1aS3BzYaUDJaOdqixzknaz,Up next: A few songs,Spotify,0,16.0,33465.0,NaN,Daily Wellness,single,0.07


In [46]:
df = df[df["artist_name"] != "Spotify"].copy()

In [47]:
df.shape

(8775, 16)

In [48]:
print("Duplicate track IDs:", df["track_id"].duplicated().sum())

Duplicate track IDs: 0


In [49]:
df["explicit"].value_counts(dropna=False)

explicit
False    6584
True     2191
Name: count, dtype: int64

In [50]:
df["album_type"].value_counts(dropna=False)

album_type
album          6022
single         2245
compilation     508
Name: count, dtype: int64

In [51]:
df["track_number"].describe()

count    8775.000000
mean        5.750427
std         6.018619
min         1.000000
25%         1.000000
50%         4.000000
75%         9.000000
max       102.000000
Name: track_number, dtype: float64

In [52]:
df[df["track_number"] > 50][
    [
        "track_name",
        "artist_name",
        "track_number",
        "album_name",
        "album_total_tracks",
        "album_type"
    ]
].sort_values("track_number", ascending=False)

,track_name,artist_name,track_number,album_name,album_total_tracks,album_type
3674,I Don't Want This Night To End,Luke Bryan,102,Best Country Of The Decade: 2011-2020,136,compilation
8336,Aguas da Amazonia: Japurá River,Philip Glass,70,Landscapes: Minimalism,100,compilation
3258,Zombie,The Cranberries,66,The Best 90s Album In The World...Ever!,77,compilation
0,3,Britney Spears,57,The Singles Collection,58,compilation
6145,Another Round For Everyone,Marcin Przybyłowicz,54,The Witcher 3: Wild Hunt (Original Game Soundt...,55,album
7682,Shaun of the Dead Main Theme (Instrumental Mix),Halloween All-Stars,52,The House of Hallowe'en - Classic Halloween Ho...,60,compilation


In [53]:
invalid_track_numbers = df[
    df["track_number"] > df["album_total_tracks"]
]

print("Invalid track numbers:", len(invalid_track_numbers))

Invalid track numbers: 0


In [54]:
print("Track popularity:", df["track_popularity"].min(), "to", df["track_popularity"].max())
print("Artist popularity:", df["artist_popularity"].min(), "to", df["artist_popularity"].max())

Track popularity: 0 to 100
Artist popularity: 0.0 to 100.0


In [55]:
print("Minimum followers:", df["artist_followers"].min())
print("Negative followers:", (df["artist_followers"] < 0).sum())

Minimum followers: 0.0
Negative followers: 0


In [56]:
df.info()

<class 'pandas.DataFrame'>
Index: 8775 entries, 0 to 8777
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   track_id            8775 non-null   str    
 1   track_name          8775 non-null   str    
 2   track_number        8775 non-null   int64  
 3   track_popularity    8775 non-null   int64  
 4   explicit            8775 non-null   bool   
 5   artist_name         8773 non-null   str    
 6   artist_popularity   8772 non-null   float64
 7   artist_followers    8772 non-null   float64
 8   artist_genres       4283 non-null   str    
 9   album_id            8775 non-null   str    
 10  album_name          8775 non-null   str    
 11  album_release_date  8775 non-null   str    
 12  album_total_tracks  8775 non-null   int64  
 13  album_type          8775 non-null   str    
 14  track_duration_min  8775 non-null   float64
 15  release_year        8775 non-null   int64  
dtypes: bool(1), float64(3)

In [57]:
df = df.reset_index(drop=True)

In [58]:
df.tail()

,track_id,track_name,track_number,track_popularity,explicit,artist_name,artist_popularity,artist_followers,artist_genres,album_id,album_name,album_release_date,album_total_tracks,album_type,track_duration_min,release_year
8770,58VTP6KnZs12PcAj5rMJ4W,the fruits,1,65,False,Paris Paloma,66.0,786890.0,NaN,48ssTnnFOCIP17sRaFNqu1,the fruits,2022-04-29,1,single,3.49,2022
8771,2dRfUOeTVpr4V4PmmqGkpm,Dark World Theme (The Legend of Zelda),4,2,False,Zelda,8.0,3263.0,NaN,0JJAfVZtn3EvZnXSQrpLSh,The Legend of Zelda,2023-08-19,7,album,1.89,2023
8772,37F0uwRSrdzkBiuj0D5UHI,Reminder,4,81,True,The Weeknd,94.0,113039308.0,NaN,2ODvWsOgouMbaA5xf0RkJe,Starboy,2016-11-25,18,album,3.65,2016
8773,42qNWdLKCI41S4uzfamhFM,Duvet,1,84,False,bôa,68.0,1188508.0,NaN,7sGYAV0xv7ZfAMzIpMl8m1,Twilight,2001-03-21,14,album,3.39,2001
8774,7KZJnMtwu2s80jO5Frthap,So Much Better,4,33,True,Eminem,88.0,104584370.0,"rap, hip hop",3vOgbDjgsZBAPwV2M3bNOj,The Marshall Mathers LP2,2013-11-05,16,album,4.36,2013


In [59]:
df.isnull().sum()

track_id                 0
track_name               0
track_number             0
track_popularity         0
explicit                 0
artist_name              2
artist_popularity        3
artist_followers         3
artist_genres         4492
album_id                 0
album_name               0
album_release_date       0
album_total_tracks       0
album_type               0
track_duration_min       0
release_year             0
dtype: int64

In [60]:
print("Duplicate rows:", df.duplicated().sum())
print("Duplicate track IDs:", df["track_id"].duplicated().sum())

Duplicate rows: 0
Duplicate track IDs: 0


In [61]:
df.to_csv(CLEAN_DATA / "spotify_final.csv", index=False)

In [62]:
print((CLEAN_DATA / "spotify_final.csv").exists())

True


In [63]:
df_final = pd.read_csv(CLEAN_DATA / "spotify_final.csv")

In [64]:
print("Final dataset shape:", df_final.shape)

Final dataset shape: (8775, 16)


In [65]:
print("Duplicate rows:", df_final.duplicated().sum())
print("Duplicate track IDs:", df_final["track_id"].duplicated().sum())
print("Missing values:")
print(df_final.isnull().sum())

Duplicate rows: 0
Duplicate track IDs: 0
Missing values:
track_id                 0
track_name               0
track_number             0
track_popularity         0
explicit                 0
artist_name              2
artist_popularity        3
artist_followers         3
artist_genres         4492
album_id                 0
album_name               0
album_release_date       0
album_total_tracks       0
album_type               0
track_duration_min       0
release_year             0
dtype: int64


In [66]:
mysql_file = CLEAN_DATA / "spotify_mysql.csv"

df_final.to_csv(
    mysql_file,
    index=False,
    encoding="utf-8-sig"
)

print(mysql_file)

/Users/jannoelvero/Documents/spotify_project/data/clean/spotify_mysql.csv


### Duplicate Track Handling

Duplicate track observations were identified during data cleaning. These records are intentionally retained because the same track may appear more than once in the dataset due to different album, single, compilation, or release contexts.

Removing these observations at this stage could reduce potentially useful information and would also change the distribution of the dataset, including the popularity threshold used later in the analysis.

However, retaining duplicate tracks creates a potential data leakage issue during machine learning. If a standard random train test split is used, different observations of the same track could appear in both the training and test sets. This could result in an overly optimistic estimate of model performance because the model would be evaluated on tracks that are also represented in the training data.

To prevent this, the machine learning stage will use a grouped train test split. Observations belonging to the same track will be assigned to the same group so that all versions of a track appear exclusively in either the training set or the test set.

**Decision:** Duplicate track observations are retained in the cleaned dataset and will be controlled through grouped splitting during machine learning.


### Investigation of Repeated Tracks

The initial duplicate check found no fully duplicated rows and no duplicated Spotify `track_id` values. However, the same recording may still appear under different Spotify track IDs when it is released as a single, album track, compilation track, or re release.

To investigate this possibility, repeated combinations of track name and artist name are examined separately. These observations are not automatically removed because they may represent legitimate release variations. The results will be used to determine whether grouped splitting is required during machine learning.


In [67]:
repeated_tracks = df[
    df.duplicated(
        subset=["track_name", "artist_name"],
        keep=False
    )
].sort_values(["track_name", "artist_name"])

print("Rows with repeated track and artist combinations:", len(repeated_tracks))
print(
    "Unique repeated track and artist combinations:",
    repeated_tracks[["track_name", "artist_name"]]
    .drop_duplicates()
    .shape[0]
)

Rows with repeated track and artist combinations: 1548
Unique repeated track and artist combinations: 697


In [68]:
repeated_tracks[
    [
        "track_id",
        "track_name",
        "artist_name",
        "album_name",
        "album_type",
        "album_release_date",
        "track_popularity"
    ]
].head(20)

,track_id,track_name,artist_name,album_name,album_type,album_release_date,track_popularity
4614,7uX3gufAoBVqKVoD3dBLLD,&burn (with Vince Staples),Billie Eilish,&burn (with Vince Staples),single,2017-12-15,56
5941,6X7qcg95CAeJpTjtg1wooM,&burn (with Vince Staples),Billie Eilish,dont smile at me,single,2017-12-22,59
5950,05od2qm2MTSKCHxy1GBp5W,12 to 12,sombr,I Barely Know Her,album,2025-08-22,89
7594,6MrLkXsMmHaYt680fhJUAq,12 to 12,sombr,12 to 12,single,2025-07-24,85
200,2ULNeSomDxVNmdDy8VxEBU,22,Taylor Swift,Red,album,2012-10-22,58
4946,2P5gfMtCcKP0CkmgqFMyPg,22,Taylor Swift,reputation Stadium Tour Surprise Song Playlist,album,2017-11-09,33
6868,3WC5CVAahvn98hiseoIvbw,22,Taylor Swift,Red (Deluxe Edition),album,2012-10-22,59
6934,5ybJm6GczjQOgTqmJ0BomP,22,Taylor Swift,Red,album,2012-10-22,0
8427,6ce41JVa6KFaryfmC2bu2W,22,Taylor Swift,Red (Big Machine Radio Release Special),album,2012-10-22,64
1959,5kqIPrATaCc2LqxVWzQGbk,7 Years,Lukas Graham,Lukas Graham,album,2016-04-01,83


### Duplicate Track Decision

The dataset contains no duplicated `track_id` values and no fully duplicated rows. However, further investigation identified 1,548 observations belonging to 697 repeated track and artist combinations.

These observations represent cases where the same track appears under different Spotify track IDs, such as album, single, deluxe edition, compilation, or other release versions. Track popularity may also differ between these versions.

These observations will not be removed because they contain legitimate release level information that may be useful for the analysis. Instead, a track grouping identifier will be created using the track name and artist name.

During machine learning, a grouped train test split will be used so that all observations belonging to the same track and artist combination are assigned entirely to either the training set or the test set. This prevents different releases of the same song from appearing in both datasets and reduces the risk of data leakage.

**Decision:** Retain repeated track observations and control them through grouped train test splitting during machine learning.


In [69]:
df["track_group"] = (
    df["track_name"].str.lower().str.strip()
    + " | "
    + df["artist_name"].str.lower().str.strip()
)

In [70]:
print("Total rows:", len(df))
print("Unique track IDs:", df["track_id"].nunique())
print("Unique track groups:", df["track_group"].nunique())

Total rows: 8775
Unique track IDs: 8775
Unique track groups: 7918


In [71]:
print("Final shape:", df.shape)
print("track_group saved:", "track_group" in df.columns)

Final shape: (8775, 17)
track_group saved: True


In [72]:
df.to_csv(CLEAN_DATA / "spotify_final.csv", index=False)

In [73]:
df_check = pd.read_csv(CLEAN_DATA / "spotify_final.csv")

print("Saved shape:", df_check.shape)
print("track_group in saved file:", "track_group" in df_check.columns)

Saved shape: (8775, 17)
track_group in saved file: True


The saved spotify_final.csv now contains the track_group variable, so the grouped train test split can be implemented later without changing Notebook 02 again.

Concern 1 status

Decision: Keep repeated releases.

Solution: Group observations by track_name + artist_name.

ML requirement: All observations from the same track_group must go entirely into either training or testing.